cmd/service/executor.go

```diff
@@ -0,0 +1,196 @@
+	// Execute the request locally if we can.
+	if request.Command.NoRemote ||
+		!request.Command.NoLocal {
+		return cmds.NewExecutor(request.Root), nil
+	}
+
+	// Everything else connects as a client.
+	var tryLaunching bool
+	serviceMaddr, addrWasProvided, err := fscmds.GetServiceMaddr(request)
+	if err != nil {
+		if !errors.Is(err, fscmds.ErrServiceNotFound) {
+			return nil, err
+		}
+		// When no arguments are provided, try auto connecting to
+		// and/or launching a service instance.
+		tryLaunching = true
+	}
+
+	if !addrWasProvided &&
+		tryLaunching {
+		pid, err := relaunchSelfAsService(request)
+		if err != nil {
+			return nil, err
+		}
+		// XXX: Don't look at this, it's only used in our _test package.
+		if pidPtr, ok := environment.(*int); ok {
+			*pidPtr = *pid
+		}
+		serviceMaddr, _, err = fscmds.GetServiceMaddr(request)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	clientHost, clientOpts, err := parseCmdsClientOptions(serviceMaddr)
+	if err != nil {
+		return nil, err
+	}
+
+	return cmdshttp.NewClient(clientHost, clientOpts...), nil
+}
+
+func parseCmdsClientOptions(maddr multiaddr.Multiaddr) (clientHost string, clientOpts []cmdshttp.ClientOpt, err error) {
+	network, dialHost, err := manet.DialArgs(maddr)
+	if err != nil {
+		return "", nil, err
+	}
+	switch network {
+	case "tcp", "tcp4", "tcp6":
+		clientHost = dialHost
+	case "unix":
+		// TODO: Consider patching cmds-lib.
+		// We want to use the URL scheme "http+unix".
+		// As-is, it prefixes the value to be parsed by pkg `url` as "http://http+unix://".
+		clientHost = fmt.Sprintf("http://%s-%s", fscmds.ServiceName, fscmds.ServerName)
+		netDialer := new(net.Dialer)
+		clientOpts = append(clientOpts, cmdshttp.ClientWithHTTPClient(&http.Client{
+			Transport: &http.Transport{
+				DialContext: func(ctx context.Context, _, _ string) (net.Conn, error) {
+					return netDialer.DialContext(ctx, network, dialHost)
+				},
+			},
+		}))
+	default:
+		return "", nil, fmt.Errorf("unsupported API address: %s", maddr)
+	}
+	return clientHost, clientOpts, nil
+}
+
+func relaunchSelfAsService(request *cmds.Request) (*int, error) {
+	// Initialize subprocess arguments and environment.
+	// Surprise, we're argv[0].
+	self, err := os.Executable()
+	if err != nil {
+		return nil, err
+	}
+	cwd, err := os.Getwd()
+	if err != nil {
+		return nil, err
+	}
+
+	// AutoExit value from request or a default.
+	serviceStopAfter, provided, err := fscmds.GetDurationArgument(request, fscmds.AutoExitParameter)
+	if err != nil {
+		return nil, err
+	}
+	if !provided {
+		serviceStopAfter = stopGrace
+	}
+
+	cmd := exec.Command(self, Name,
+		fmt.Sprintf("--%s=%s", fscmds.AutoExitParameter.Name, serviceStopAfter),
+	)
+	cmd.Dir = cwd
+	cmd.Env = os.Environ()
+
+	// Setup IPC
+	servicePipe, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	// Issue the command.
+	if err = cmd.Start(); err != nil {
+		return nil, err
+	}
+	proc := cmd.Process
+	procState := cmd.ProcessState
```

Comment:
Even though these are just shorthand variables. They should stand out more than hide. The indented var draws attention to the area near the call to `Start` which is significant, but easy to gloss over when cluttered like this.
Besides, we use the multi-line var liberally throughout the code, so something something consistency too.
```suggestion
var (
		proc      = cmd.Process
		procState = cmd.ProcessState
	)
```

Reply:
On second thought, it's probably better to at least move procState into the error encountered branch where it's used. I'll reconsider this whole section later.